go/tools/gopackagesdriver/flatpackage.go

```diff
@@ -0,0 +1,134 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"go/parser"
+	"go/token"
+	"os"
+	"strconv"
+	"strings"
+)
+
+type ResolvePkgFunc func(importPath string) *FlatPackage
+
+// Copy and pasted from golang.org/x/tools/go/packages
+type FlatPackagesError struct {
+	Pos  string // "file:line:col" or "file:line" or "" or "-"
+	Msg  string
+	Kind FlatPackagesErrorKind
+}
+
+type FlatPackagesErrorKind int
+
+const (
+	UnknownError FlatPackagesErrorKind = iota
+	ListError
+	ParseError
+	TypeError
+)
+
+func (err FlatPackagesError) Error() string {
+	pos := err.Pos
+	if pos == "" {
+		pos = "-" // like token.Position{}.String()
+	}
+	return pos + ": " + err.Msg
+}
+
+// FlatPackage is the JSON form of Package
+// It drops all the type and syntax fields, and transforms the Imports
+type FlatPackage struct {
+	ID              string
+	Name            string              `json:",omitempty"`
+	PkgPath         string              `json:",omitempty"`
+	Errors          []FlatPackagesError `json:",omitempty"`
+	GoFiles         []string            `json:",omitempty"`
+	CompiledGoFiles []string            `json:",omitempty"`
+	OtherFiles      []string            `json:",omitempty"`
+	ExportFile      string              `json:",omitempty"`
+	Imports         map[string]string   `json:",omitempty"`
+	Standard        bool                `json:",omitempty"`
+}
+
+type PackageFunc func(pkg *FlatPackage)
+type PathResolverFunc func(path string) string
+
+func resolvePathsInPlace(prf PathResolverFunc, paths []string) {
+	for i, path := range paths {
+		paths[i] = prf(path)
+	}
+}
+
+func WalkFlatPackagesFromJSON(jsonFile string, onPkg PackageFunc) error {
```

Comment:
Could this just return a list of `FlatPackage` instead of accepting a callback? I think that would be a little more readable.

Reply:
it's to avoid rewalking the list over and over. on a side note, `yield` would have been perfect.